File: src/examples/messages/java/src/main/java/hu/gds/examples/messages/MessageHeader.java
package hu.gds.examples.messages;

import org.msgpack.core.MessageBufferPacker;
import org.msgpack.core.MessageUnpacker;

import java.io.IOException;

import static hu.gds.examples.messages.MessagePackUtil.*;

public class MessageHeader {

    /*
        [
            "user", --> 'user name'
            "793ed37a-a30c-44cb-848b-ad30c1c52358", --> 'message id'
            'System.currentTimeMillis()', --> 'create time'
            'System.currentTimeMillis()', --> 'request time'
            false, --> 'is fragmented'
            null, --> 'first fragment'
            null, --> 'last fragment'
            null, --> 'offset'
            null, --> 'full data size'
            '[0..12]', --> 'data type'
            '...' --> 'data'
        ]
     */
    public static void packHeader(MessageBufferPacker packer, int dataType) throws IOException {
        //user name
        packer.packString("user");

        //message id
        packer.packString("793ed37a-a30c-44cb-848b-ad30c1c52358");

        //create time
        packer.packLong(System.currentTimeMillis());

        //request time
        packer.packLong(System.currentTimeMillis());

        //is fragmented
        packer.packBoolean(false);

        //first fragment
        packer.packNil();

        //last fragment
        packer.packNil();

        //offset
        packer.packNil();

        //full data size
        packer.packNil();

        //data type
        packer.packInt(dataType);
    }

    public static int unpackHeader(MessageUnpacker unpacker) throws IOException {
        //user name
        String userName = unpacker.unpackString();

        //message id
        String messageId = unpacker.unpackString();

        //create time
        Long createTime = unpacker.unpackLong();

        //request time
        Long requestTime = unpacker.unpackLong();

        //is fragmented
        Boolean isFragmented = unpacker.unpackBoolean();

        //first fragment
        Boolean firstFragment = unPackBoolean(unpacker);

        //last fragment
        Boolean lastFragment = unPackBoolean(unpacker);

        //offset
        Integer offset = unpackInteger(unpacker);

        //full data size
        Integer fullDataSize = unpackInteger(unpacker);

        //dataType
        Integer dataType = unpacker.unpackInt();

        return dataType;
    }
}

File: src/examples/messages/c#/messages/Message/Header/MessageHeader.cs
using MsgPack;
using MsgPack.Serialization;
using System;
using System.Collections.Generic;
using System.Text;

namespace messages
{
    public class MessageHeader : IPackable, IUnpackable
    {
        private string user;
        private string messageId;
        private long createTime;
        private long requestTime;
        private bool isFragmented;
        private bool? firstFragment;
        private bool? lastFragment;
        private int? offset;
        private int? fullDataSize;
        private int dataType;

        public MessageHeader(string user, string messageId, long createTime, long requestTime,
            bool isFragmented, bool? firstFragment, bool? lastFragment, int? offset,
            int? fullDataSize, int dataType)
        {
            this.user = user;
            this.messageId = messageId;
            this.createTime = createTime;
            this.requestTime = requestTime;
            this.isFragmented = isFragmented;
            this.firstFragment = firstFragment;
            this.lastFragment = lastFragment;
            this.offset = offset;
            this.fullDataSize = fullDataSize;
            this.dataType = dataType;
        }

        internal MessageHeader() {}

        public string User => user;
        public string MessageId => messageId;
        public long CreateTime => createTime;
        public long RequestTime => requestTime;
        public bool IsFragmented => isFragmented;
        public bool? FirstFragment => firstFragment;
        public bool? LastFragment => lastFragment;
        public int? Offset => offset;
        public int? FullDataSize => fullDataSize;
        public int DataType => dataType;

        public void PackToMessage(Packer packer, PackingOptions options)
        {
            packer.Pack(user);
            packer.Pack(messageId);
            packer.Pack(createTime);
            packer.Pack(requestTime);
            packer.Pack(isFragmented);
            packer.Pack(firstFragment);
            packer.Pack(lastFragment);
            packer.Pack(offset);
            packer.Pack(fullDataSize);
            packer.Pack(dataType);
        }

        public void UnpackFromMessage(Unpacker unpacker)
        {
            unpacker.ReadString(out user);
            unpacker.ReadString(out messageId);
            unpacker.ReadInt64(out createTime);
            unpacker.ReadInt64(out requestTime);
            unpacker.ReadBoolean(out isFragmented);
            unpacker.ReadNullableBoolean(out firstFragment);
            unpacker.ReadNullableBoolean(out lastFragment);
            unpacker.ReadNullableInt32(out offset);
            unpacker.ReadNullableInt32(out fullDataSize);
            unpacker.ReadInt32(out dataType);
        }
    }
}
